### ISSUE TITLE
`torch` / `lightning` integration

### ISSUE THREAD

**Issue by fkiraly** [SimonBlanke/Hyperactive#195]
It would be nice to have integration with basic `torch` and `lightning` tuning workflows, to allow autoML style tuning.

This would require a `BaseExperiment` descendant class `TorchExperiment` which takes a `DataLoader`, a `LightningModule`, and a `Trainer`, performs a training run, and returns the validation loss/score.

Since tuning is not API preserving in `torch` / `lightning`, I would suggest an additional function that produces parameters for a tuned network, or the initialized tuned network with said parameters.

Comments appreciated on the design.

**Comment by fkiraly** [comment]
If it works in this generality, it should be tested whether it can be used to replace or extend the `pytorch-forecasting` tuning feature - which is currently `optuna` based.

**Comment by fkiraly** [comment]
I would suggest, implement the torch experiment first, the rest can build on that later.

**Comment by amitsubhashchejara** [comment]
Sure, I'll start!

**Comment by amitsubhashchejara** [comment]
Hello @fkiraly, I have a question. To perform a training run and return a validation loss/score, we need two dataloaders, a train_loader and a validation_loader. So should the `TorchExperiment` class accept two dataloaders? Also, which score should be returned along with the loss? Accuracy, AUCROC, F1?

**Comment by amitsubhashchejara** [comment]
Maybe the class can accept a `datamodule`, or both `train_loader` and `cv_loader` should be provided. For the score, we can set `val_loss` as the default.

**Comment by fkiraly** [comment]
Yes, I was thinking of using a `lightning` `DataModule`, which bundles both training and validation loader.

I would advice, let´s start with `lightning` based parametrization, we can later think about whether we would also like a `torch` native one.

**Comment by amitsubhashchejara** [comment]
Okay thankyou! There is one more point: should the trainer run on the original `LightningModule` or on a copy of the same? Because we want to save the `params` of the model, the user has to use the `save_hyperparameters` in the `__init__` of the  `LightningModule`, only then can we create a new instance in the `_paramnames` function with the same params.

**Comment by fkiraly** [comment]
> Because we want to save the `params` of the model

Could you kindly explain why we want that, e.g., what in the design implies this?

**Comment by amitsubhashchejara** [comment]
```
def _paramnames(self):
        """Return the parameter names of the search.

        Returns
        -------
        list of str
            The parameter names of the search parameters.
        """
        return list(self.estimator.get_params().keys())
```
This is a function in the template

**Comment by fkiraly** [comment]
I see - no need to copy the objects then, I think the return of `_paramnames_` should simply be the parameters that the user could set in the model.

**Comment by amitsubhashchejara** [comment]
Okay, thank you so much!

**Comment by agobbifbk** [comment]
Nice one!

**Comment by amitsubhashchejara** [comment]
@fkiraly and others, please check!
